File: code/java/ch12/src/main/java/com/github/zxh0/luago/state/LuaTable.java
package com.github.zxh0.luago.state;

import com.github.zxh0.luago.number.LuaNumber;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class LuaTable {

    LuaTable metatable;
    private List<Object> arr;
    private Map<Object, Object> map;
    // used by next()
    private Map<Object, Object> keys;
    private Object lastKey;
    private boolean changed;

    LuaTable(int nArr, int nRec) {
        if (nArr > 0) {
            arr = new ArrayList<>(nArr);
        }
        if (nRec > 0) {
            map = new HashMap<>(nRec);
        }
    }

    boolean hasMetafield(String fieldName) {
        return metatable != null && metatable.get(fieldName) != null;
    }

    int length() {
        return arr == null ? 0 : arr.size();
    }

    Object get(Object key) {
        key = floatToInteger(key);

        if (arr != null && key instanceof Long) {
            int idx = ((Long) key).intValue();
            if (idx >= 1 && idx <= arr.size()) {
                return arr.get(idx - 1);
            }
        }

        return map != null ? map.get(key) : null;
    }

    void put(Object key, Object val) {
        if (key == null) {
            throw new RuntimeException("table index is nil!");
        }
        if (key instanceof Double && ((Double) key).isNaN()) {
            throw new RuntimeException("table index is NaN!");
        }

        changed = true;
        key = floatToInteger(key);
        if (key instanceof Long) {
            int idx = ((Long) key).intValue();
            if (idx >= 1) {
                if (arr == null) {
                    arr = new ArrayList<>();
                }

                int arrLen = arr.size();
                if (idx <= arrLen) {
                    arr.set(idx - 1, val);
                    if (idx == arrLen && val == null) {
                        shrinkArray();
                    }
                    return;
                }
                if (idx == arrLen + 1) {
                    if (map != null) {
                        map.remove(key);
                    }
                    if (val != null) {
                        arr.add(val);
                        expandArray();
                    }
                    return;
                }
            }
        }

        if (val != null) {
            if (map == null) {
                map = new HashMap<>();
            }
            map.put(key, val);
        } else {
            if (map != null) {
                map.remove(key);
            }
        }
    }

    private Object floatToInteger(Object key) {
        if (key instanceof Double) {
            Double f = (Double) key;
            if (LuaNumber.isInteger(f)) {
                return f.longValue();
            }
        }
        return key;
    }

    private void shrinkArray() {
        for (int i = arr.size() - 1; i >= 0; i--) {
            if (arr.get(i) == null) {
                arr.remove(i);
            }
        }
    }

    private void expandArray() {
        if (map != null) {
            for (int idx = arr.size() + 1; ; idx++) {
                Object val = map.remove((long) idx);
                if (val != null) {
                    arr.add(val);
                } else {
                    break;
                }
            }
        }
    }

    Object nextKey(Object key) {
        if (keys == null || (key == null && changed)) {
            initKeys();
            changed = false;
        }

        Object nextKey = keys.get(key);
        if (nextKey == null && key != null && key != lastKey) {
            throw new RuntimeException("invalid key to 'next'");
        }

        return nextKey;
    }

    private void initKeys() {
        if (keys == null) {
            keys = new HashMap<>();
        } else {
            keys.clear();
        }
        Object key = null;
        if (arr != null) {
            for (int i = 0; i < arr.size(); i++) {
                if (arr.get(i) != null) {
                    long nextKey = i + 1;
                    keys.put(key, nextKey);
                    key = nextKey;
                }
            }
        }
        if (map != null) {
            for (Object k : map.keySet()) {
                Object v = map.get(k);
                if (v != null) {
                    keys.put(key, k);
                    key = k;
                }
            }
        }
        lastKey = key;
    }

}

File: code/go/ch13/src/luago/state/lua_table.go
package state

import "math"
import "luago/number"

type luaTable struct {
	metatable *luaTable
	arr       []luaValue
	_map      map[luaValue]luaValue
	keys      map[luaValue]luaValue // used by next()
	lastKey   luaValue              // used by next()
	changed   bool                  // used by next()
}

func newLuaTable(nArr, nRec int) *luaTable {
	t := &luaTable{}
	if nArr > 0 {
		t.arr = make([]luaValue, 0, nArr)
	}
	if nRec > 0 {
		t._map = make(map[luaValue]luaValue, nRec)
	}
	return t
}

func (self *luaTable) hasMetafield(fieldName string) bool {
	return self.metatable != nil &&
		self.metatable.get(fieldName) != nil
}

func (self *luaTable) len() int {
	return len(self.arr)
}

func (self *luaTable) get(key luaValue) luaValue {
	key = _floatToInteger(key)
	if idx, ok := key.(int64); ok {
		if idx >= 1 && idx <= int64(len(self.arr)) {
			return self.arr[idx-1]
		}
	}
	return self._map[key]
}

func _floatToInteger(key luaValue) luaValue {
	if f, ok := key.(float64); ok {
		if i, ok := number.FloatToInteger(f); ok {
			return i
		}
	}
	return key
}

func (self *luaTable) put(key, val luaValue) {
	if key == nil {
		panic("table index is nil!")
	}
	if f, ok := key.(float64); ok && math.IsNaN(f) {
		panic("table index is NaN!")
	}

	self.changed = true
	key = _floatToInteger(key)
	if idx, ok := key.(int64); ok && idx >= 1 {
		arrLen := int64(len(self.arr))
		if idx <= arrLen {
			self.arr[idx-1] = val
			if idx == arrLen && val == nil {
				self._shrinkArray()
			}
			return
		}
		if idx == arrLen+1 {
			delete(self._map, key)
			if val != nil {
				self.arr = append(self.arr, val)
				self._expandArray()
			}
			return
		}
	}
	if val != nil {
		if self._map == nil {
			self._map = make(map[luaValue]luaValue, 8)
		}
		self._map[key] = val
	} else {
		delete(self._map, key)
	}
}

func (self *luaTable) _shrinkArray() {
	for i := len(self.arr) - 1; i >= 0; i-- {
		if self.arr[i] == nil {
			self.arr = self.arr[0:i]
		}
	}
}

func (self *luaTable) _expandArray() {
	for idx := int64(len(self.arr)) + 1; true; idx++ {
		if val, found := self._map[idx]; found {
			delete(self._map, idx)
			self.arr = append(self.arr, val)
		} else {
			break
		}
	}
}

func (self *luaTable) nextKey(key luaValue) luaValue {
	if self.keys == nil || (key == nil && self.changed) {
		self.initKeys()
		self.changed = false
	}

	nextKey := self.keys[key]
	if nextKey == nil && key != nil && key != self.lastKey {
		panic("invalid key to 'next'")
	}

	return nextKey
}

func (self *luaTable) initKeys() {
	self.keys = make(map[luaValue]luaValue)
	var key luaValue = nil
	for i, v := range self.arr {
		if v != nil {
			self.keys[key] = int64(i + 1)
			key = int64(i + 1)
		}
	}
	for k, v := range self._map {
		if v != nil {
			self.keys[key] = k
			key = k
		}
	}
	self.lastKey = key
}
